src/indicators/keltner_channel.rs

```diff
@@ -0,0 +1,202 @@
+use std::fmt;
+
+use crate::errors::*;
+use crate::indicators::{AverageTrueRange, ExponentialMovingAverage};
+use crate::{Close, Next, Reset};
+
+/// Keltner Channel (KC).
+///
+/// A Keltner Channel is an indicator showing the Average True Range (ATR) of a
+/// price surrounding a central moving average. The ATR bands are typically
+/// shown 'k' times moved away from the moving average.
+///
+/// # Formula
+///
+/// See EMA, ATR documentation.
+///
+/// KC is composed as:
+///
+///  * _KC<sub>Middle Band</sub>_ - Exponential Moving Average (EMA).
+///  * _KC<sub>Upper Band</sub>_ = EMA + ATR of observation * multipler (usually 2.0)
+///  * _KC<sub>Lower Band</sub>_ = EMA - ATR of observation * multipler (usually 2.0)
+///
+/// # Example
+///
+///```
+/// use ta::indicators::{KeltnerChannel, KeltnerChannelOutput};
+/// use ta::Next;
+///
+/// let mut kc = KeltnerChannel::new(3, 2.0_f64).unwrap();
+///
+/// let out_0 = kc.next(2.0);
+///
+/// let out_1 = kc.next(5.0);
+///
+/// assert_eq!(out_0.average, 2.0);
+/// assert_eq!(out_0.upper, 2.0);
+/// assert_eq!(out_0.lower, 2.0);
+///
+/// assert_eq!(out_1.average, 3.5);
+/// assert_eq!(out_1.upper, 6.5);
+/// assert_eq!(out_1.lower, 0.5);
+/// ```
+///
+/// # Links
+///
+/// * [Keltner channel, Wikipedia](https://en.wikipedia.org/wiki/Keltner_channel)
+#[derive(Debug, Clone)]
+pub struct KeltnerChannel {
+    length: u32,
+    multiplier: f64,
+    atr: AverageTrueRange,
+    ema: ExponentialMovingAverage,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct KeltnerChannelOutput {
+    pub average: f64,
+    pub upper: f64,
+    pub lower: f64,
+}
+
+impl KeltnerChannel {
+    pub fn new(length: u32, multiplier: f64) -> Result<Self> {
+        if multiplier <= 0.0 {
+            return Err(Error::from_kind(ErrorKind::InvalidParameter));
+        }
+        Ok(Self {
+            length,
+            multiplier,
+            atr: AverageTrueRange::new(length)?,
+            ema: ExponentialMovingAverage::new(length)?,
+        })
+    }
+
+    pub fn length(&self) -> u32 {
+        self.length
+    }
+
+    pub fn multiplier(&self) -> f64 {
+        self.multiplier
+    }
+}
+
+impl Next<f64> for KeltnerChannel {
+    type Output = KeltnerChannelOutput;
+
+    fn next(&mut self, input: f64) -> Self::Output {
+        let atr = self.atr.next(input);
+        let average = self.ema.next(input);
+
+        Self::Output {
+            average,
+            upper: average + atr * self.multiplier,
+            lower: average - atr * self.multiplier,
+        }
+    }
+}
+
+impl<'a, T: Close> Next<&'a T> for KeltnerChannel {
```

Comment:
I think here, according to the formula in wikipedia, we should use `Close`, `High`, `Low`.

It's a rare indicator, where implementation of `next<f64>` and `next<T>` are different.
For `next<T>` we'd like:
* `ema.next((input.close + input.high + input.low) / 3.0)`
* `atr.next(input)`

Note: ATR has also 2 alternative implementations for `next<f64>`  and `next<T>`